### PR TITLE
🌱 cache secrets in KCP, CABPK and ClusterCacheTracker

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:
This PR ensures we cache secrets in KCP, CABPK and the ClusterCacheTracker (this should cover all read calls to the "ca" and kubeconfig certificates.

Based on our scale test data the uncached secret get calls take about 50% of our KCP reconcile duration and leads to a relatively high request rate to the apiserver.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related to #8814

### CODE REVIEW

File: util/secret/certificates.go [hunk 1]
```diff
@@ -189,27 +190,39 @@ func (c Certificates) GetByPurpose(purpose Purpose) *Certificate {
 }
 
 // Lookup looks up each certificate from secrets and populates the certificate with the secret data.
-func (c Certificates) Lookup(ctx context.Context, ctrlclient client.Client, clusterName client.ObjectKey) error {
+func (c Certificates) Lookup(ctx context.Context, secretCachingClient, ctrlclient client.Client, clusterName client.ObjectKey) error {
```
Comment (sbueringer) [inline util/secret/certificates.go:193]: We have to decide if we want to just make this breaking change or basically copy & paste this function and LookupOrGenerate.

I think usually nobody almost nobody is using this util and it's not too hard to adopt. But I'm fine with going through the normal deprecation

@fabriziopandini @killianmuldoon @ykakarap  opinions?

**Comment by sbueringer** [comment]
/assign @fabriziopandini @killianmuldoon @ykakarap

/hold
for
/test pull-cluster-api-e2e-full-main

**Comment by sbueringer** [comment]
/hold cancel
CI is green

File: util/secret/certificates.go (hunk 1 above)
Comment (fabriziopandini) [reply util/secret/certificates.go:193]: I will go through deprecation, we are after the first beta date and we can make an alias without this paramater + handle this being null

Comment (sbueringer) [reply util/secret/certificates.go:193]: Hm the UX becomes a bit awkward when folks are supposed to pass in a nil client in some cases

Comment (sbueringer) [reply util/secret/certificates.go]: Pushed a new version, WDYT? (+/- needs more godoc if we go with that variant)

File: internal/controllers/topology/cluster/patches/engine.go [hunk 2]
```diff
@@ -268,7 +268,7 @@ func createRequest(blueprint *scope.ClusterBlueprint, desired *scope.ClusterStat
 	// has its own state, e.g. version or replicas. This state is used to calculate builtin variables,
 	// which can then be used e.g. to compute the machine image for a specific Kubernetes version.
 	for mdTopologyName, md := range desired.MachineDeployments {
-		// Lookup MachineDeploymentTopology definition from cluster.spec.topology.
+		// LookupCached MachineDeploymentTopology definition from cluster.spec.topology.
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/controllers/topology/cluster/patches/engine.go:271]: Yup Intellij :/ Probably missed to review when it asked if those changes were intended

**Review by fabriziopandini — COMMENTED** [review]
/lgtm
/approve

/hold
for squash
feel free to unhold

**Pull request merged**